toolsrc/src/tls12-download.c

```diff
@@ -0,0 +1,311 @@
+#include <Windows.h>
+#include <process.h>
+#include <winhttp.h>
+/*
+ * This program must be as small as possible, because it is committed in binary form to the
+ * vcpkg github repo to enable downloading the main vcpkg program on Windows 7, where TLS 1.2 is
+ * unavailable to PowerShell.
+ * To that end it avoids using C runtime functions (beyond the vcruntime ones the compiler
+ * injects itself).
+ * (In testing as of 2021-01-07, this version that doesn't link with the CRT is ~8kb, whereas a
+ * hello world program that does link with the CRT is ~300kb)
+ */
+
+static void __declspec(noreturn) win32_abort()
+{
+    /*
+     * Note that TerminateProcess does not return when called from the terminated process, see
+     * https://github.com/MicrosoftDocs/sdk-api/pull/626
+     */
+    TerminateProcess(GetCurrentProcess(), 3);
+}
+
+static size_t wide_length(const wchar_t* str)
+{
+    size_t answer = 0;
+    while (*str)
+    {
+        ++answer;
+        ++str;
+    }
+    return answer;
+}
+
+static void write_message(const HANDLE std_out, const wchar_t* msg)
+{
+    size_t wchars_to_write = wide_length(msg);
+    if (wchars_to_write == 0)
+    {
+        return;
+    }
+
+    if (wchars_to_write > 65535)
+    {
+        win32_abort();
+    }
+
+    if (WriteConsoleW(std_out, msg, wchars_to_write, 0, 0))
+    {
+        return;
+    }
+
+    // this happens if output has been redirected
+    int narrow_chars = WideCharToMultiByte(CP_ACP, 0, msg, (int)wchars_to_write, 0, 0, 0, 0);
+    if (narrow_chars == 0)
+    {
+        win32_abort();
+    }
+
+    char* narrow_buffer = HeapAlloc(GetProcessHeap(), 0, (size_t)narrow_chars);
+    if (WideCharToMultiByte(CP_ACP, 0, msg, (int)wchars_to_write, narrow_buffer, narrow_chars, 0, 0) == 0)
+    {
+        win32_abort();
+    }
+
+    while (narrow_chars != 0)
+    {
+        DWORD chars_written;
+        if (!WriteFile(std_out, narrow_buffer, (DWORD)narrow_chars, &chars_written, 0))
+        {
+            win32_abort();
+        }
+
+        narrow_chars -= (int)chars_written;
+    }
+
+    if (!HeapFree(GetProcessHeap(), 0, narrow_buffer))
+    {
+        win32_abort();
+    }
+}
+
+static void write_number(const HANDLE std_out, DWORD number)
+{
+    wchar_t buffer[11]; // 4294967295\0
+    wchar_t* first_digit = buffer + 11;
+    *--first_digit = L'\0';
+    if (number == 0)
+    {
+        *--first_digit = L'0';
+    }
+    else
+    {
+        do
+        {
+            *--first_digit = L'0' + number % 10;
+            number /= 10;
+        } while (number != 0);
+    }
+
+    write_message(std_out, first_digit);
+}
+
+static void write_hex(const HANDLE std_out, DWORD number)
+{
+    wchar_t buffer[] = L"0x00000000";
+    wchar_t* first_digit = buffer + (sizeof(buffer) / sizeof(wchar_t)) - 1;
+    while (number != 0)
+    {
+        *--first_digit = L"0123456789ABCDEF"[number % 16];
+        number /= 16;
+    }
```

Comment:
```suggestion
    const size_t nybble_size = sizeof(DWORD) * 2;
    for (int i = 0; i < nybble_size; ++i)
    {
        const int shift = 4 * (nybble_size - i - 1);
        buffer[i + 2] = L"0123456789ABCDEF"[(number >> shift) & 0xF];
    }
```
I prefer this method, but up to you.